src/strategies/Staking.sol

```diff
@@ -4,174 +4,348 @@ pragma solidity =0.8.16;
-            amount: debtToken.balanceOf(address(this)),
-            interestRateMode: 2,
-            onBehalfOf: address(this)
-        });
-        AAVE.withdraw({asset: address(WSTETH), amount: aToken.balanceOf(address(this)), to: address(this)});
-
-        // Swap wstETH to WETH
-        ISwapRouter.ExactInputSingleParams memory params = ISwapRouter.ExactInputSingleParams({
-            tokenIn: address(WSTETH),
-            tokenOut: address(WETH),
-            fee: 500,
-            recipient: address(this),
-            deadline: block.timestamp,
-            amountIn: WSTETH.balanceOf(address(this)),
-            amountOutMinimum: 0, // TODO: slippage protection
-            sqrtPriceLimitX96: 0
-        });
-        ROUTER.exactInputSingle(params);
-
-        // Pay back flash loan using WETH
-        WETH.safeTransfer(address(vault), amount);
-
-        lastPosPnl = int256(WETH.balanceOf(address(this))) - int256(lastPosSize);
-
-        // Unwrap all WETH
-        IWETH(address(WETH)).withdraw(WETH.balanceOf(address(this)));
+    enum Loan {
+        open,
+        close
     }
-
+  
     function receiveFlashLoan(
         ERC20[] memory tokens,
         uint256[] memory amounts,
         uint256[] memory, /* feeAmounts */
         bytes memory userData
     ) external override {
-        require(msg.sender == address(vault), "Staking: only vault");
+        require(msg.sender == address(balancer), "Staking: only balancer");
+
+        uint256 ethBorrowed;
+        uint daiBorrowed;
+        (Loan l) = abi.decode(userData, (Loan));
+
+        if (l == Loan.close) {
+            // We only flashloan dai in certain cases during divestment
+            if (tokens.length == 2) {
+                daiBorrowed = amounts[0];
+                ethBorrowed = amounts[1];
+            } else {
+                ethBorrowed = amounts[0];
+            }
+            _endPosition(ethBorrowed, daiBorrowed);
+        } else {
+            ethBorrowed = amounts[0];
+            _openPosition(ethBorrowed);
+        }
+        
+        // Payback loan
+        WETH.safeTransfer(address(balancer), ethBorrowed);
+        // We take slightly more from the user than they want to deposit (for slippage). If there's any left over send back to the user.
+        if (l == Loan.open) WETH.safeTransfer(address(vault), WETH.balanceOf(address(this)));
+
+        // Repay Dai loan
+        if (daiBorrowed > 0) DAI.safeTransfer(address(balancer), daiBorrowed);
+    }
+ 
+    function _openPosition(uint ethBorrowed) internal {
+        // Convert WETH to stEth
+        IWETH(address(WETH)).withdraw(ethBorrowed);
+        Address.sendValue(payable(address(LIDO)), ethBorrowed);
+
+        uint compoundDebtEth = _openCdp(ERC20(address(LIDO)).balanceOf(address(this)));
+
+        // Convert eth to weth in order to pay back balancer flash loan of weth
+        IWETH(address(WETH)).deposit{value: compoundDebtEth}();
+    }
+
+    uint public cdpId;
```

Comment:
Can be constant state variable. or We can put inside function.

Reply:
It can't be constant since we don't know it until the cdp is opened. 

Since opening the cdp doesn't actually require any capital, we can actually do it inside of the constructor though! This way both the cdpId and the urn address can be immutable.